tests/unit/test_charm.py

```diff
@@ -3,25 +3,187 @@
 #
 # Learn more about testing at: https://juju.is/docs/sdk/testing
 """Unit tests module."""
+from typing import List
 import unittest
-from unittest.mock import patch
+from unittest.mock import patch, Mock
+import yaml
 
-from ops.testing import Harness
+from ops.model import StatusBase
+from ops.testing import Harness, ActionFailed
 
 from src.charm import OperatorMachineCharm
 
 APP_NAME = "canonical-livepatch-server"
 
 
+# pylint: disable=too-many-public-methods
 class TestCharm(unittest.TestCase):
     """Unit tests class."""
 
     def setUp(self):
         self.harness = Harness(OperatorMachineCharm)
         self.addCleanup(self.harness.cleanup)
         self.harness.disable_hooks()
+
+        self.snap_mock, snap_cache_patch = self.new_mock_snap_cache()
+        snap_cache_patch.start()
+        self.addCleanup(snap_cache_patch.stop)
+
         self.harness.begin()
 
+    def new_mock_snap_cache(self):
+        """create a mock snap cache"""
+        mock = Mock()
+        snap_cache = {"canonical-livepatch-server": mock}
+        return mock, patch("src.charm.SnapCache", return_value=snap_cache)
+
+    # wokeignore:rule=master
+    def test_legacy_db_master_changed(self):
+        """test `_legacy_db_master_changed event` handler."""
+        rel_id = self.harness.add_relation("livepatch", "livepatch")
+        self.harness.add_relation_unit(rel_id, f"{APP_NAME}/1")
+        self.harness.set_leader(True)
+        self.harness.enable_hooks()
+
+        legacy_db_rel_id = self.harness.add_relation("database-legacy", "postgres")
+
+        # The `ops-lib-pgsql` library calls `leader-get` and `leader-set` tools
+        # from juju help-tools, so we need to mock calls that try to spawn a
+        # subprocess.
+        stored_data = "'{}'"
+
+        def set_database_name_using_juju_leader_set(cmd: List[str]):
+            nonlocal stored_data
```

Comment:
What about returning a value instead of the use of the global? Or is that not possible because of how we use it in the mock?

Reply:
Yeah, it's beacuse of the mocked functions expected behaviour, and of course, Python's way of handling scopes.